crates/wasmparser/src/validator/operators.rs

```diff
@@ -1488,6 +1559,83 @@ where
+                    match self.label_types(block.0, block.1)?.last() {
+                        Some(ValType::Ref(rt)) if rt.is_concrete_type_ref() => {
+                            let new_cont_id = rt.type_index().unwrap().unpack().as_core_type_id().expect("expected canonicalized index");
+                            let new_func_ty = self.func_type_of_cont_type_at(Either::B(new_cont_id))?;
+                            // Check that (ts2' -> ts2) <: $ft
+                            if new_func_ty.params().len() != tag_ty.results().len() || !self.is_subtype_many(0, new_func_ty.params(), 0, tag_ty.results())
+                                || old_func_ty.results().len() != new_func_ty.results().len() || !self.is_subtype_many(0, old_func_ty.results(), 0, new_func_ty.results()) {
+                                bail!(self.offset, "type mismatch in continuation type")
+                            }
+                            let expected_nargs = tag_ty.params().len() + 1;
+                            let actual_nargs = self
+                                .label_types(block.0, block.1)?
+                                .len();
+                            if actual_nargs != expected_nargs {
+                                bail!(self.offset, "type mismatch: expected {expected_nargs} label result(s), but label is annotated with {actual_nargs} results")
+                            }
+
+                            let labeltys = self
+                                .label_types(block.0, block.1)?
+                                .take(expected_nargs - 1);
+
+                            // Check that ts1'' <: ts1'.
+                            for (tagty, &lblty) in labeltys.zip(tag_ty.params()) {
+                                if !self.resources.is_subtype(lblty, tagty) {
+                                    bail!(self.offset, "type mismatch between tag type and label type")
+                                }
+                            }
+                        }
+                        Some(ty) => {
+                            bail!(self.offset, "type mismatch: {}", ty_to_str(ty))
+                        }
+                        _ => bail!(self.offset,
+                                   "type mismatch: instruction requires continuation reference type but label has none")
+                    }
+                }
+                Handle::OnSwitch { tag } => {
+                    let tag_ty = self.tag_at(tag)?;
+                    if tag_ty.params().len() != 0 {
+                        bail!(self.offset, "type mismatch: non-empty tag parameter type")
+                    }
+                }
+            }
+        }
+        Ok(old_func_ty)
+    }
+
+    /// Applies `is_subtype` pointwise two equally sized collections
+    /// (i.e. equally sized after skipped elements).
+    fn is_subtype_many(
+        &mut self,
+        skipcnt1: usize,
+        ts1: &[ValType],
+        skipcnt2: usize,
+        ts2: &[ValType],
```

Comment:
Instead of taking a `skipcntN` could this only take two slices? It should be possible to chop off the front of the slices in callers explicitly if necessary

Reply:
Ah yes, it is much prettier. Thanks!